### PR TITLE
Add integration test for rolling update

### PR DESCRIPTION
**Pull request by dongluochen**
This test validates the procedure of a successful rolling update. 

Ping @tonistiigi @dperny. 

Signed-off-by: Dong Chen dongluo.chen@docker.com

### CODE REVIEW

File: integration-cli/daemon.go [hunk 1]
```diff
@@ -527,3 +527,76 @@ func (d *Daemon) checkActiveContainerCount(c *check.C) (interface{}, check.Comme
 	}
 	return len(strings.Split(strings.TrimSpace(out), "\n")), check.Commentf("output: %q", string(out))
 }
+
+func (d *Daemon) checkRunningTaskCount(c *check.C) (interface{}, check.CommentInterface) {
+	// get service list
+	serviceOutput, err := d.Cmd("service", "ls", "-q")
+	c.Assert(err, checker.IsNil)
+	if len(strings.TrimSpace(serviceOutput)) == 0 {
+		return 0, nil
```
Comment (LK4D4) [inline integration-cli/daemon.go]: shouldn't we return comment here as well?

File: integration-cli/daemon.go [hunk 2]
```diff
@@ -527,3 +527,76 @@ func (d *Daemon) checkActiveContainerCount(c *check.C) (interface{}, check.Comme
 	}
 	return len(strings.Split(strings.TrimSpace(out), "\n")), check.Commentf("output: %q", string(out))
 }
+
+func (d *Daemon) checkRunningTaskCount(c *check.C) (interface{}, check.CommentInterface) {
+	// get service list
+	serviceOutput, err := d.Cmd("service", "ls", "-q")
```
Comment (cpuguy83) [inline integration-cli/daemon.go]: I think it would likely be better to hit the tasks API endpoint and iterate over the tasks directly.

**Comment by mlaventure** [comment]
Failure on windows seems genuine

```
# github.com/docker/distribution/registry/api/errcode
vendor\src\github.com\docker\distribution\registry\api\errcode\register.go:74: undefined: http.StatusTooManyRequests
```

**Comment by dperny** [comment]
i got the same windows failure in #24563

**Comment by dperny** [comment]
Ok I deleted my comment but then in my email there is actually a comment about #24563 failing on windows for the same reason. Now it passes, no changes made. Windows build may be flaky.

**Comment by tonistiigi** [comment]
That win2lin ci box has wrong version of go. @mlaventure removed it.

File: integration-cli/daemon.go (hunk 2 above)
Comment (tonistiigi) [inline integration-cli/daemon.go]: Agree. This is also used by the api tests.

File: integration-cli/docker_api_swarm_test.go [hunk 3]
```diff
@@ -371,6 +371,67 @@ func (s *DockerSwarmSuite) TestApiSwarmServicesCreateGlobal(c *check.C) {
...
+	d3 := s.AddDaemon(c, true, false)
+	time.Sleep(1 * time.Second) // make sure all daemons are ready to accept tasks
+
+	// create a different tag
+	out, err := d1.Cmd("tag", "busybox:latest", "busybox:test")
+	c.Assert(err, checker.IsNil, check.Commentf(out))
+	out, err = d2.Cmd("tag", "busybox:latest", "busybox:test")
+	c.Assert(err, checker.IsNil, check.Commentf(out))
+	out, err = d3.Cmd("tag", "busybox:latest", "busybox:test")
+	c.Assert(err, checker.IsNil, check.Commentf(out))
```
Comment (tonistiigi) [inline integration-cli/docker_api_swarm_test.go]: nit: for loop seems better here, `AddDaemon` can also be in there.

File: integration-cli/daemon.go (hunk 2 above)
Comment (dongluochen) [inline integration-cli/daemon.go]: Will do.

File: integration-cli/daemon.go (hunk 1 above)
Comment (dongluochen) [inline integration-cli/daemon.go]: this means no services. I can add that as a comment.

File: integration-cli/docker_api_swarm_test.go (hunk 3 above)
Comment (dongluochen) [inline integration-cli/docker_api_swarm_test.go]: will do.

**Comment by dongluochen** [comment]
Several changes: 
1. Use API instead of CLI to get tasks. 
2. Do not validate `update-delay` as it's hard to estimate run time in different systems. Also `--update-delay` is not meant to be accurate. 
3. Validate each batch should change `update-parallelism` number of tasks.

**Comment by icecrime** [comment]
Removing from milestone (it's not strictly part of the release!).

**Comment by dongluochen** [comment]
Thank @tonistiigi for pointing out I can compare map in Go. The code is much simpler now.

File: integration-cli/docker_api_swarm_test.go [hunk 4]
```diff
@@ -371,6 +371,52 @@ func (s *DockerSwarmSuite) TestApiSwarmServicesCreateGlobal(c *check.C) {
 	waitAndAssert(c, defaultReconciliationTimeout, d5.checkActiveContainerCount, checker.Equals, 1)
 }
 
+func (s *DockerSwarmSuite) TestApiSwarmServicesUpdate(c *check.C) {
+	const nodeCount = 3
+	var daemons [nodeCount]*SwarmDaemon
+	daemons[0] = s.AddDaemon(c, true, true)
+	for i := 1; i < nodeCount; i++ {
+		daemons[i] = s.AddDaemon(c, true, false)
```
Comment (tonistiigi) [inline integration-cli/docker_api_swarm_test.go]: nit: `daemons[i] = s.AddDaemon(c, true, i == 0)`

**Comment by tonistiigi** [comment]
LGTM

File: integration-cli/docker_api_swarm_test.go [hunk 5]
```diff
@@ -371,6 +371,52 @@ func (s *DockerSwarmSuite) TestApiSwarmServicesCreateGlobal(c *check.C) {
...
+	}
+	time.Sleep(1 * time.Second) // make sure all daemons are ready to accept tasks
```
Comment (cpuguy83) [inline integration-cli/docker_api_swarm_test.go]: Nit, do for loop on each daemon checking for `d.Status.State == swarm.NodeStateReady`

**Comment by thaJeztah** [comment]
ping @cpuguy83 looks like the PR is. updated PTAL

**Comment by cpuguy83** [comment]
LGTM

**Comment by thaJeztah** [comment]
Adding back to the milestone; it was mainly removed to not be a blocker for release

**Pull request merged**